### DISCUSSION TITLE
Init-only & required properties support

### DISCUSSION THREAD

**Discussion by smoothdeveloper** [fsharp/fslang-design#695] [Language and core library RFC discussions]
This is initialising only, the required discussion about F# compiler to honour C# types that bear `init` and `required` modifiers on properties.

* [RFC](https://github.com/fsharp/fslang-design/blob/main/FSharp-7.0/FS-1127-init-only-and-required-properties.md)
* [PR](https://github.com/dotnet/fsharp/pull/13490)

**Comment by smoothdeveloper** [comment]
Assuming C# required properties works with the "initializer" syntax (I gather that is the point of the feature); there will be need for F# to distinguish the property setter that are accepted into any method call, versus those that are for a constructor call:

```csharp
public class WithRequired {
   public required int Foo {get; set;}
   public required int Bar {get; init;}
}

var ok = new WithRequired {Foo=42, Bar =42}
```

```fsharp
type WithRequiredFactory =
  static member MakeDefault() = WithRequired(Foo=42, Bar=42)
  static member MakeWithFoo(foo) = WithRequired(Foo=foo, Bar=42)

let notOk = WithRequiredFactory.MakeDefault(Bar=42) // can't set because `Bar` is init-only
let notOk2 = WithRequiredFactory.MakeWithFoo(1, Bar=42) // can't set because `Bar` is init-only
let ok = WithRequiredFactory.MakeDefault(Foo=43) // an extra call to Foo setter is appended before returning the expression
let ok2 = WithRequired(Foo=42, Bar=42)
```

This may have weird interaction with `inline` member.

Initially, I always seen initializer syntax in C# as "syntax sugar" only, but now it seems to evolve into "checks on the caller/initializer".

**Reply by NinoFloris** [reply]
Hahah what is this syntax even. I'm extremely surprised this is valid today. Was it a deliberate choice for this to work outside constructor calls @dsyme?

**Reply by vzarytovskii** [reply]
> Hahah what is this syntax even. I'm extremely surprised this is valid today. Was it a deliberate choice for this to work outside constructor calls @dsyme?

Yeah, it's pretty much a property initializer syntax outside of constructor. I have learned it today after chatting with @dsyme :D

**Reply by vzarytovskii** [reply]
> Assuming C# required properties works with the "initializer" syntax (I gather that is the point of the feature); there will be need for F# to distinguish the property setter that are accepted into any method call, versus those that are for a constructor call:
> 
> ```csharp
> public class WithRequired {
>    public required int Foo {get; set;}
>    public required int Bar {get; init;}
> }
> 
> var ok = new WithRequired {Foo=42, Bar =42}
> ```
> 
> ```fsharp
> type WithRequiredFactory =
>   static member MakeDefault() = WithRequired(Foo=42, Bar=42)
>   static member MakeWithFoo(foo) = WithRequired(Foo=foo, Bar=42)
> 
> let notOk = WithRequiredFactory.MakeDefault(Bar=42) // can't set because `Bar` is init-only
> let notOk2 = WithRequiredFactory.MakeWithFoo(1, Bar=42) // can't set because `Bar` is init-only
> let ok = WithRequiredFactory.MakeDefault(Foo=43) // an extra call to Foo setter is appended before returning the expression
> let ok2 = WithRequired(Foo=42, Bar=42)
> ```
> 
> This may have weird interaction with `inline` member.
> 
> Initially, I always seen initializer syntax in C# as "syntax sugar" only, but now it seems to evolve into "checks on the caller/initializer".

Yeah, this is one of the points in the "TODO" list on the PR, to make sure we only allow to call init-only setter from the constructor. I've learned of this possibility today :D

**Reply by smoothdeveloper** [reply]
Syntax is great syntactical sugar when doing "object programming", I hope it doesn't disappear, albeit, I wish it used `:` rather than `=`.

In the F# 2.0 language specification:

> 14.4 Method Application Resolution
> Resolves post-hoc property assignments.

There is also similar note in *6.4.1 Basic Application Expressions*

**Comment by 64J0** [comment]
Hey, is it possible to annotate an F#'s property value with this **required** modifier? Similar to that C# snippet mentioned at this discussion:

> ```csharp
> public class WithRequired {
>    public required int Foo {get; set;}
>    public required int Bar {get; init;}
> }
> 
> var ok = new WithRequired {Foo=42, Bar =42}
> ```

**Reply by vzarytovskii** [reply]
I don't think it is.

**Reply by 64J0** [reply]
Got it, thanks for the clarification @vzarytovskii
